refactor: choose setup or subscription mode to create a subscription

nirinaHelena opened the pull request.

nirinaHelena committed:
refactor: use setUp mode

nirinaHelena committed:
refactor: put trialPeriodDays to 0 and eligibleFrom to now in SubscriptionServiceIT::initiate_subscription

nirinaHelena committed:
chore: format

nirinaHelena committed:
refactor: use Subscription or SetUP mode depends on the subscription create date

nirinaHelena committed:
test: subscriptionService initiate (wip)

nirinaHelena committed:
refactor: put StripeSession into a class

hei-ryan commented on src/main/java/app/bpartners/api/service/subscription/StripeSessionFactory.java:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  @SneakyThrows
+  public void simulateSubscriptionScheduleCreation(
```
Why simulate ?

hei-ryan commented on src/main/java/app/bpartners/api/service/subscription/StripeSessionFactory.java:
```diff
@@ -0,0 +1,152 @@
+            .build());
+  }
+
+  public Session createSessionSetUp(
```
Must be private, as it is only called inside the Factory. 

Change the unit test to handle it as well.

hei-ryan commented on src/main/java/app/bpartners/api/service/subscription/SubscriptionService.java:
```diff
@@ -377,6 +350,40 @@ public Redirection initiateSubscription(
+      SubscriptionProduct subscriptionProduct)
+      throws StripeException {
+    if (trialEnd.isAfter(temporalUtils.fifthOfNextMonth())) {
+      return stripeSessionFactory.createSessionSetUp(
```
No. 

Actually, it must be reveresed, where is the test I've asked for to look at it correctly ?

If it's the 6th April, we can directly create the subscription as the period between 6th April to 5th May is less than natural_billing_cycle (30 days). But when the period before the next billing is > 30 day that's what you should handle.

nirinaHelena committed:
refactor: put createSession in StripeSessionFactory.java

nirinaHelena committed:
test: StripeSessionFactoryTest::create_session_if_trial_end_is_before_next_billing_date

nirinaHelena committed:
test: StripeSessionFactoryIT::create_session_if_trial_end_is_after_next_billing_date
StripeSessionFactoryIT::create_session_if_trial_end_is_before_next_billing_date

nirinaHelena committed:
chore: format
